jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java

```diff
@@ -854,4 +850,35 @@ public void remove()
             };
         }
     }
+
+    /**
+     * Support for Jetty `jetty-loader:` protocol.
+     *
+     * <p>
+     * Internally uses {@link Loader#getResource(String)} to find URL
+     * </p>
+     */
+    public static class JettyLoaderURLStreamHandlerProvider extends URLStreamHandlerProvider
+    {
+        @Override
+        public URLStreamHandler createURLStreamHandler(String protocol)
+        {
+            if ("jetty-loader".equals(protocol))
+            {
+                return new URLStreamHandler()
+                {
+                    @Override
+                    protected URLConnection openConnection(URL u) throws IOException
+                    {
+                        String path = u.getPath();
+                        URL url = Loader.getResource(path);
+                        if (url == null)
+                            throw new FileNotFoundException("Unable to find Loader resource: " + path);
+                        return url.openConnection();
+                    }
+                };
+            }
+            return null;
+        }
+    }
```

Comment:
This class needs better explanation.  I'm very cautious about URL and any JVM-wide classes like this.

Reply:
Unlike pre-Java 9 days, a URL protocol handler is now on-demand, and not registered permanently across the entire JVM once and only once.

In the pre-java 9 days the to handle url protocols you had 2 techniques.

1. A System Property that needed to be set at start of the JVM (`-Djava.protocol.handler.pkgs=<package-name>`)
2. Using the static method [`java.net.URL.setURLStreamHandlerFactory()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/URL.html#setURLStreamHandlerFactory(java.net.URLStreamHandlerFactory)), which could  ever be called once.

Now, the protocol handlers are service loader based, and only are searched when a new protocol is encountered during an `openStream()` event.

Reply:
Hmmm maybe.... but what problem are we solving by switching to catalogs?  They are still baked into the jar, so we are just switching from java to xml that still has to be built.

So why do we need to introduce the extra complexity of a URL handler?  What is the benefit?

Reply:
The way it exists now in Java 9+, this logic could even be isolated to only operate in a thread that is tagged by the `XmlParser`.

Eg:

* parse() - tag thread
* only resolve when tag is present on current thread
* done with parse() - tag removed
* all other uses are just a null reference